pandas/core/dtypes/cast.py

```diff
@@ -1026,7 +1026,7 @@ def maybe_cast_to_datetime(value, dtype, errors="raise"):
             )
 
             if is_datetime64 and not is_dtype_equal(dtype, _NS_DTYPE):
-                if dtype.name in ("datetime64", "datetime64[ns]"):
+                if dtype.name in ("datetime64", "datetime64[ns]", "datetime64[D]"):
```

Comment:
just use ```dtype.kind == 'M'```

Reply:
Since `pandas` doesn't support to convert [`datetime[ps]`](https://github.com/pandas-dev/pandas/blob/master/pandas/tests/series/test_constructors.py#L1352), I used `(dtype.kind == 'M') and (dtype.name != 'datetime64[ps]')`